### PR TITLE
Custom search (html form update required) + RSS + Atom

### PR DESCRIPTION
**Pull request by imaginal**
+ rss / atom feed in serp
+ new fields in es: index_card, post.actual_region, post.post_type, estate.region
+ new manage.py command reset_large_tasks for fix large spotter tasks
+ new manage.py command ftsearch_index_card for create index_card
+ upd manage.py command loadnacp add index_card, actual_region, post_type, estate.region
* html search form need improvement
* aren't set selected in form inputs

### CODE REVIEW

File: declarations_site/catalog/management/commands/ftsearch_index_card.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                "general.post.office",
+                "general.post.region",
+                "intro.declaration_year",
+                "intro.doc_type",
+                "declaration.source",
+                "declaration.url"
+            ]
+            index_card = ""
+
+            for path in card_fields:
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/ftsearch_index_card.py]: We have dpath for that.

File: declarations_site/catalog/management/commands/ftsearch_index_card.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+from elasticsearch_dsl import Mapping, Text
+from django.core.management.base import BaseCommand
+from catalog.elastic_models import Declaration
+
+
+class Command(BaseCommand):
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/ftsearch_index_card.py]: Can you merge that command with ftsearch_and_apostrophes.py?

File: declarations_site/catalog/management/commands/link_corrected.py [hunk 3]
```diff
@@ -258,7 +258,7 @@ def handle(self, *args, **options):
     def save_to_excel(self, fname):
         workbook = xlsxwriter.Workbook(fname)
         hits_worksheet = workbook.add_worksheet()
-        mediocre_worksheet = workbook.add_worksheet() 
+        mediocre_worksheet = workbook.add_worksheet()
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/link_corrected.py:261]: So nice of you

File: declarations_site/catalog/management/commands/loadnacp.py [hunk 4]
```diff
@@ -331,10 +374,33 @@ def _parse_me(cls, base_fname):
...
+        # if set only one region use it value for second one
+        if not resp["general"]["post"]["actual_region"] and resp["general"]["post"]["region"]:
+            resp["general"]["post"]["actual_region"] = resp["general"]["post"]["region"]
+        elif not resp["general"]["post"]["region"] and resp["general"]["post"]["actual_region"]:
+            resp["general"]["post"]["region"] = resp["general"]["post"]["actual_region"]
+
+        # concatenate some fields for library index card
+        index_card = ""
+
+        for path in cls.card_fields:
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/loadnacp.py]: Move that to helpers?

File: declarations_site/catalog/management/commands/loadnacp.py [hunk 5]
```diff
@@ -331,10 +374,33 @@ def _parse_me(cls, base_fname):
...
+            head = resp
+            for p in path.split('.'):
+                head = head.get(p, "")
+                if not head:
+                    break
+            index_card += str(head or "") + " "
+
+        resp["index_card"] = index_card.strip()
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/loadnacp.py]: I bet, ES doesn't give a flying fuck about trailing whitespaces, but ok.

File: declarations_site/spotter/views.py [hunk 6]
```diff
@@ -61,17 +57,41 @@ def save_search(request):
         return redirect('search_list')
 
     task = SearchTask(user=request.user, query=query, deepsearch=deepsearch)
+    task.query_params = query_params
     task.save()
 
-    first_run(task)
-    if not send_newtask_notify(task):
+    if not first_run(task):
+        messages.warning(request, 'Не вдалось створити завдання, спробуйте спростити '+
```
Comment (dchaplinsky) [inline declarations_site/spotter/views.py]: Plus is redundant here, afaik.

Comment (dchaplinsky) [reply declarations_site/spotter/views.py]: Also, you are missing space before the plus sign.

File: declarations_site/declarations_site/settings.py [hunk 7]
```diff
@@ -138,6 +138,10 @@
 FROM_EMAIL = 'robot@declarations.com.ua'
 EMAIL_HOST = 'localhost'
 
+RSS_AUTHOR_NAME = 'Сайт «Декларації» - проект канцелярської сотні'
```
Comment (dchaplinsky) [inline declarations_site/declarations_site/settings.py]: Why in lowercase?

File: declarations_site/declarations_site/settings.py [hunk 8]
```diff
@@ -138,6 +138,10 @@
...
+RSS_AUTHOR_LINK = 'https://declarations.com.ua'
+RSS_AUTHOR_EMAIL = 'monti.czardas@gmail.com'
```
Comment (dchaplinsky) [inline declarations_site/declarations_site/settings.py]: Can you use any @declarations.com.ua email instead?

Comment (dchaplinsky) [reply declarations_site/declarations_site/settings.py]: For example dbihus@declarations.com.ua

File: declarations_site/spotter/management/commands/reset_large_tasks.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+from spotter.models import SearchTask, TaskReport
+from spotter.utils import first_run
+
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    help = 'Reset spotter tasks with more that 1K results'
```
Comment (dchaplinsky) [inline declarations_site/spotter/management/commands/reset_large_tasks.py:13]: What for?

File: declarations_site/spotter/management/commands/reset_large_tasks.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    def handle(self, *args, **options):
+        for task in SearchTask.objects.filter(is_enabled=True, is_deleted=False,
+                                              found_total__gt=1000):
+            try:
+                task_reports = TaskReport.objects.filter(task=task.id)
+                print('Task "{}" total found {} total reports {}'.format(
```
Comment (dchaplinsky) [inline declarations_site/spotter/management/commands/reset_large_tasks.py]: Please use self.stdout/stderr.write

File: declarations_site/catalog/rss.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def populate_feed(request, data, feed_type):
+    # now available only for search results
+    if 'query' not in data:
+        raise Http404()
+
+    current_site = get_current_site(request)
+
+    link = request.get_full_path().replace('format=', 'from_format=')
```
Comment (dchaplinsky) [inline declarations_site/catalog/rss.py]: Potentially unsafe, it's better to parse/unparse it.

**Review by dchaplinsky — COMMENTED** [review]
First part of comments, TBC

File: declarations_site/catalog/static/js/main.js [hunk 12]
```diff
@@ -118,6 +154,16 @@ $(function() {
             }
         });
 
+        //add two blocks to TOC
+        if( $('#similar_by_sirname').length > 0 ) {
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/main.js]: Surname

File: declarations_site/catalog/static/js/main.js [hunk 13]
```diff
@@ -118,6 +154,16 @@ $(function() {
...
+            $('<li class="devider"></li>').appendTo('#nacp-toc ul');
```
Comment (dchaplinsky) [inline declarations_site/catalog/static/js/main.js]: divider

File: declarations_site/catalog/templates/results.jinja [hunk 14]
```diff
@@ -16,31 +18,199 @@
...
             {{ social("Результати пошуку по: %s. " % query,
-                              "http://declarations.com.ua%s" % request.path ,
-                              add_disqus=False) }}
+                              "http://declarations.com.ua%s" % request.get_full_path() ,
+                              add_disqus=False,
+                              add_rss=True) }}
         </div>
 
-        <h1 id="page-header"><span>{{ query|default("Всі", true)|escape }} ({{ results.paginator.count }})</span>
+        <div id="ex-search-form">
```
Comment (dchaplinsky) [inline declarations_site/catalog/templates/results.jinja]: Move that to macro

File: declarations_site/catalog/templates/results.jinja [hunk 15]
```diff
@@ -16,31 +18,199 @@
...
+                                    <li><a href="#" data-value="Севастополь">м. Севастополь</li>
+                                </ul>
+                            </div>
+                            <input id="search-field-region" class="linked-input" type="hidden" name="region_value" />
+
+                            <div class="form-group">
+                              <div class="radio">
+                                <label>
+                                  <input type="radio" checked="" name="region_type" value="region" id="search-field-region_type-1">
+                                  Регіон прописки
```
Comment (dchaplinsky) [inline declarations_site/catalog/templates/results.jinja]: Реєстрації

File: declarations_site/catalog/management/commands/loadnacp.py [hunk 16]
```diff
@@ -300,6 +325,24 @@ def _parse_me(cls, base_fname):
                         "relations": member.get("subjectRelation", "")
                     })
 
+        # get regions from estate list
+        if "step_3" in data and isinstance(data["step_3"], dict) and data["step_3"]:
```
Comment (dchaplinsky) [inline declarations_site/catalog/management/commands/loadnacp.py:314]: What about
4.   ОБ'ЄКТИ НЕЗАВЕРШЕНОГО БУДІВНИЦТВА

File: declarations_site/catalog/templates/results.jinja (hunk 14 above)
Comment (dchaplinsky) [reply declarations_site/catalog/templates/results.jinja]: Or include

File: declarations_site/catalog/utils.py [hunk 17]
```diff
@@ -25,6 +26,103 @@ def replace_apostrophes(s):
     return s.replace("`", "'").replace("’", "'")
 
 
+def replace_arg(request, key, value):
+    args = request.GET.copy()
+    args[key] = value
+    return args.urlencode()
+
+
+def apply_search_sorting(search, sort=""):
```
Comment (dchaplinsky) [inline declarations_site/catalog/utils.py]: What about directions of sort?

File: declarations_site/catalog/templates/results.jinja [hunk 18]
```diff
@@ -16,31 +18,199 @@
...
+                              add_disqus=False,
+                              add_rss=True) }}
         </div>
 
-        <h1 id="page-header"><span>{{ query|default("Всі", true)|escape }} ({{ results.paginator.count }})</span>
+        <div id="ex-search-form">
+            <form class="box" action="{{ url("search") }}" method="GET">
+
+                <div class="input-group search-string">
+                    <input type="text" name="q" class="form-control search-name" id="search-form" placeholder="Наприклад: Айварас Абромавичус; Суддя; Хмельницька область; Державна фіскальна служба" autocomplete="off" value="{% if query %}{{ query }}{% endif %}" />
```
Comment (dchaplinsky) [inline declarations_site/catalog/templates/results.jinja]: Let's change it:
"Наприклад: Анатолій Матіос; суддя; Луцьк; Міністерство юстиції; Rolex"

File: declarations_site/catalog/utils.py [hunk 19]
```diff
@@ -25,6 +26,103 @@ def replace_apostrophes(s):
...
+
+    query_op = {"query": query, "operator": "and"}
+    query_kw = {default_field: query_op}
+
+    search = base_search.query("match", **query_kw)
+    search = apply_search_filters(search, filters)
+
+    nwords = len(re.findall(r'\w{4,}', query))
+
+    if nwords > 2 and deepsearch and not search.count():
```
Comment (dchaplinsky) [inline declarations_site/catalog/utils.py:133]: What about old logic for shallow search?

File: declarations_site/spotter/views.py [hunk 20]
```diff
@@ -61,17 +57,41 @@ def save_search(request):
...
+def save_search(request):
+    query = replace_apostrophes(request.GET.get("q", "")).strip()
+    deepsearch = bool(request.GET.get("deepsearch", ""))
+
+    params = request.GET.copy()
+    for key in ("q", "deepsearch", "format", "page"):
+        if key in params:
+            params.pop(key)
+
+    response = do_save_search(request, query, deepsearch, params.urlencode())
```
Comment (dchaplinsky) [inline declarations_site/spotter/views.py:89]: You moved it in separate function just for the sake of readability?

**Review by dchaplinsky — COMMENTED** [review]
I'm done

File: declarations_site/catalog/management/commands/ftsearch_index_card.py (hunk 1 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/ftsearch_index_card.py]: dpath process dots notation?

File: declarations_site/catalog/utils.py (hunk 17 above)
Comment (imaginal) [reply declarations_site/catalog/utils.py]: there no direction in form now and when @aturik was ask about that nobody replied that is required

File: declarations_site/catalog/utils.py (hunk 19 above)
Comment (imaginal) [reply declarations_site/catalog/utils.py:133]: we was considered that should be changed, isn't it?

File: declarations_site/declarations_site/settings.py (hunk 7 above)
Comment (imaginal) [reply declarations_site/declarations_site/settings.py]: Suggest better

File: declarations_site/declarations_site/settings.py (hunk 8 above)
Comment (imaginal) [reply declarations_site/declarations_site/settings.py]: if it alive - of course

File: declarations_site/spotter/management/commands/reset_large_tasks.py (hunk 9 above)
Comment (imaginal) [reply declarations_site/spotter/management/commands/reset_large_tasks.py:13]: fix old crappy minimal_shold_match=2 results

File: declarations_site/spotter/views.py (hunk 6 above)
Comment (imaginal) [reply declarations_site/spotter/views.py]: are you sure about space?

File: declarations_site/spotter/views.py (hunk 20 above)
Comment (imaginal) [reply declarations_site/spotter/views.py:89]: ... and use more than one exit point, this required for ajax multiple query adding form

File: declarations_site/catalog/management/commands/ftsearch_index_card.py (hunk 2 above)
Comment (dchaplinsky) [reply declarations_site/catalog/management/commands/ftsearch_index_card.py]: Hello Wladimir

File: declarations_site/catalog/management/commands/ftsearch_index_card.py (hunk 1 above)
Comment (dchaplinsky) [reply declarations_site/catalog/management/commands/ftsearch_index_card.py]: dpath.util.values(doc, path, separator='.')

File: declarations_site/catalog/management/commands/loadnacp.py (hunk 4 above)
Comment (dchaplinsky) [reply declarations_site/catalog/management/commands/loadnacp.py]: Yes/No/Maybe?

Comment (dchaplinsky) [reply declarations_site/catalog/management/commands/loadnacp.py]: You are using same code at least in two places

File: declarations_site/spotter/views.py (hunk 6 above)
Comment (dchaplinsky) [reply declarations_site/spotter/views.py]: Well, more or less:
If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator.

File: declarations_site/declarations_site/settings.py (hunk 7 above)
Comment (dchaplinsky) [reply declarations_site/declarations_site/settings.py]: проект Канцелярської сотні

File: declarations_site/declarations_site/settings.py (hunk 8 above)
Comment (dchaplinsky) [reply declarations_site/declarations_site/settings.py]: it is, not sure if someone reading it tho.

File: declarations_site/spotter/management/commands/reset_large_tasks.py (hunk 9 above)
Comment (dchaplinsky) [reply declarations_site/spotter/management/commands/reset_large_tasks.py:13]: Ok

File: declarations_site/spotter/management/commands/reset_large_tasks.py (hunk 10 above)
Comment (dchaplinsky) [reply declarations_site/spotter/management/commands/reset_large_tasks.py]: Any decision taken?

File: declarations_site/catalog/rss.py (hunk 11 above)
Comment (dchaplinsky) [reply declarations_site/catalog/rss.py]: Any decision taken?

File: declarations_site/catalog/templates/results.jinja (hunk 14 above)
Comment (dchaplinsky) [reply declarations_site/catalog/templates/results.jinja]: Any decision taken?

File: declarations_site/catalog/management/commands/loadnacp.py (hunk 16 above)
Comment (dchaplinsky) [reply declarations_site/catalog/management/commands/loadnacp.py:314]: Hey!

File: declarations_site/catalog/utils.py (hunk 17 above)
Comment (dchaplinsky) [reply declarations_site/catalog/utils.py]: Ok

File: declarations_site/catalog/utils.py (hunk 19 above)
Comment (dchaplinsky) [reply declarations_site/catalog/utils.py:133]: What is the new logic then?

File: declarations_site/spotter/views.py (hunk 20 above)
Comment (dchaplinsky) [reply declarations_site/spotter/views.py:89]: Ok

File: declarations_site/catalog/management/commands/ftsearch_index_card.py (hunk 2 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/ftsearch_index_card.py]: Done

File: declarations_site/catalog/management/commands/ftsearch_index_card.py (hunk 1 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/ftsearch_index_card.py]: OK

File: declarations_site/catalog/management/commands/link_corrected.py (hunk 3 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/link_corrected.py:261]: You welcome

File: declarations_site/catalog/management/commands/loadnacp.py (hunk 4 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/loadnacp.py]: Ok, move to model

File: declarations_site/spotter/views.py (hunk 6 above)
Comment (imaginal) [reply declarations_site/spotter/views.py]: Oh, you mean about that one, the space will be reborned here

File: declarations_site/spotter/management/commands/reset_large_tasks.py (hunk 10 above)
Comment (imaginal) [reply declarations_site/spotter/management/commands/reset_large_tasks.py]: Let it be

File: declarations_site/catalog/rss.py (hunk 11 above)
Comment (imaginal) [reply declarations_site/catalog/rss.py]: will be QueryDict

File: declarations_site/catalog/templates/results.jinja (hunk 14 above)
Comment (imaginal) [reply declarations_site/catalog/templates/results.jinja]: What exactly?

File: declarations_site/catalog/utils.py (hunk 19 above)
Comment (imaginal) [reply declarations_site/catalog/utils.py:133]: 1. apply only if 2 or more words in query
2. minimum_should_match = num_words - 1  # or - 2 for 5 or more words

File: declarations_site/catalog/templates/results.jinja (hunk 14 above)
Comment (dchaplinsky) [reply declarations_site/catalog/templates/results.jinja]: You have a huge chunk of code (which is basically search form) sitting in the template of search results. My proposition here is to move that form completely to the include or macro

File: declarations_site/spotter/management/commands/reset_large_tasks.py (hunk 10 above)
Comment (dchaplinsky) [reply declarations_site/spotter/management/commands/reset_large_tasks.py]: Let it be what?

File: declarations_site/catalog/management/commands/loadnacp.py (hunk 16 above)
Comment (imaginal) [reply declarations_site/catalog/management/commands/loadnacp.py:314]: YEH

**Pull request merged**